src/v/s3/signature.h

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include "http/client.h"
+#include "outcome.h"
+#include "units.h"
+#include "utils/named_type.h"
+#include "vassert.h"
+
+#include <chrono>
+#include <seastarx.h>
+#include <string_view>
+
+namespace s3 {
+
+using aws_region_name = named_type<ss::sstring, struct s3_aws_region_name>;
+using public_key_str = named_type<ss::sstring, struct s3_public_key_str>;
+using private_key_str = named_type<ss::sstring, struct s3_private_key_str>;
+using timestamp = std::chrono::time_point<std::chrono::system_clock>;
+
+/// Time source for the signature_v4.
+/// Can be used to get and format current time or to
+/// format the pre-defined time for testing.
+class time_source {
+    enum {
+        formatted_date_len = 9,      // format is 20201231\0
+        formatted_datetime_len = 17, // format is 20201231T123100Z\0
+    };
+
+public:
+    /// \brief Initialize time-source
+    /// Defult time-source uses std::chrono::system_clock.
+    time_source();
+
+    /// \brief Initialize time-source using the hardcoded
+    /// value for testing.
+    ///
+    /// \param instant is a timestamp that time_source should use
+    explicit time_source(timestamp instant);
+
+    time_source(time_source const&) = delete;
+    time_source& operator=(time_source const&) = delete;
+    time_source(time_source&&) = default;
+    time_source& operator=(time_source&&) = default;
+    ~time_source() = default;
```

Comment:
noexcept

Reply:
i think the idea is that we want to std::terminate on an exception in the dtor, right @senior7515 ?

Reply:
right, makes sense, will fix that